### PR TITLE
Blocks: move bootstrapped block types to Redux state

### PR DESCRIPTION
**Pull request by jsnajdr**
This is a spinoff from the async block loading prototype (#51778), and it moves the `serverSideBlockDefinitions` to Redux state of the `core/blocks` store, and adds some private actions and selectors to that store to work with the bootstrapped block types.

The `serverSideBlockDefinitions` object is now in `state.bootstrappedBlockTypes`, and almost all code from `unstable__bootstrapServerSideBlockDefinitions` has been moved to the `boostrappedBlockTypes` reducer. It's the code that adds some missing properties and converts property names to camel case. There is a new `addBootstrappedBlock` private action to add boostrapped blocks, and `unstable__bootstrapServerSideBlockDefinitions` calls this action in a loop.

The `__experimentalRegisterBlockType` public action has been renamed to a private action `addUnprocessedBlock`. It's very unlikely that it's used by any plugin. A [wpdirectory.net search](https://wpdirectory.net/search/01H8440XF64KMV4FCG17903QZK) shows two matches, but on closer inspection both of them just bundle the `@wordpress/blocks` package.

Similarly, the `__experimentalGetUnprocessedBlockTypes` public selector has been renamed to private `getUnprocessedBlockTypes`.

Now block registration works like this:
1. Server-bootstrapped block metadata are added with `addBootstrappedBlock` and stored in `state.bootstrappedBlockTypes`.
2. Block settings passed to `wp.blocks.registerBlockTypes` are stored in `state.unprocessedBlockTypes`.
3. There is a `processBlockType` function (now extracted to a separate module) that takes these two pieces of information ("bootstrapped" metadata and "unprocessed" settings), merges them together, fills default values, and runs the result through `registerBlockType` filters. The filtered processed result is stored in `state.blockTypes`. That's where all the user code, like `wp.blocks.getBlockTypes()`, gets the block types from.
4. There continues to be the `__experimentalReapplyBlockTypeFilters` action, called after editor initialization, that reruns the `processBlockType` job and re-creates `state.blockTypes` once again from the raw data. That accounts for filters that are registered late.

A good followup for this PR would be a separate PR that renames and stabilizes the `unstable__bootstrapServerSideBlockDefinitions` and `__experimentalReapplyBlockTypeFilters` APIs. Both of them are by now established parts of any Gutenberg editor initialization.

### CODE REVIEW

**Review by tyxla — COMMENTED** [review]
@jsnajdr haven't done a thorough review just yet, but in the meantime, I noticed that the failing e2e tests have a consistent kind of "The block "undefined" must have a title." error and it persists after retrying. Seems to be something related to the changes suggested here.

**Comment by jsnajdr** [comment]
> I noticed that the failing e2e tests have a consistent kind of "The block "undefined" must have a title." error

That was a bug in the `ADD_UNPROCESSED_BLOCK_TYPE` reducer, resulting in block name being registered as `undefined`. The `__experimentalReapplyBlockTypeFilters` action then didn't work at all. I pushed a fix for the reducer, and will now work also on a unit test for `__experimentalReapplyBlockTypeFilters`. It has no coverage at all.

**Comment by gziolo** [comment]
> > I noticed that the failing e2e tests have a consistent kind of "The block "undefined" must have a title." error
> 
> That was a bug in the `ADD_UNPROCESSED_BLOCK_TYPE` reducer, resulting in block name being registered as `undefined`. The `__experimentalReapplyBlockTypeFilters` action then didn't work at all. I pushed a fix for the reducer, and will now work also on a unit test for `__experimentalReapplyBlockTypeFilters`. It has no coverage at all.

See https://github.com/WordPress/gutenberg/pull/34299 for reference. The experimental action had indirect e2e test coverage added, but it would be great to add more tests to increase confidence in this logic. I think the issue is that the `name` is no longer passed as part of the `settings` object to the action. Here is the old code that covers it:

https://github.com/WordPress/gutenberg/blob/44a8678975ef85b53199e8161f554479ca1041c0/packages/blocks/src/api/registration.js#L299L315

I believe that the modified action gets raw settings now, while in the past it would standardize it with default values first.

**Comment by gziolo** [comment]
Thank you for opening this PR. It's definitely the area of the codebase that should be improved after we confirmed over time that https://github.com/WordPress/gutenberg/pull/34299 resolved the issue with the order of applying filters for the block registration. I need to spend more time processing the changes proposed. I'd like to share some higher-level thoughts for now after I left a previous comment regarding the failing unit test.

> A good followup for this PR would be a separate PR that renames and stabilizes the unstable__bootstrapServerSideBlockDefinitions  and __experimentalReapplyBlockTypeFilters APIs. Both of them are by now established parts of any Gutenberg editor initialization.

An important note is that
`unstable__bootstrapServerSideBlockDefinitions` was never meant to be the part of public API. Back then when we had to ship WP 5.0, we didn't have a better way to control what gets exposed to the client for block types registered on the server. Today, we should refactor code to use REST API endpoint for block types instead as outlined in https://github.com/WordPress/gutenberg/issues/22812. My only concern would be whether there are any legacy concerns as the post editor is initialized on the server with the context of a specific post type and a post id, so blocks in theory could get modified during registration with filters based on that information. The same issue doesn't apply to the site editor though.

**Comment by jsnajdr** [comment]
> it would be great to add more tests to increase confidence in this logic.

Yes, I added a unit tests that tests the "reapply block filters" logic. Before, it was tested only through e2e tests.

The bug was a simple mistake, I was reading the `name` value from a value where it wasn't.

> I believe that the modified action gets raw settings now, while in the past it would standardize it with default values first.

Yes, I decided to rearrangle the logic in such a way that `bootstrappedBlockTypes` and `unprocessedBlockTypes` now store the raw, unmodified data that are passed to `registerBlockType( metadata, settings )`. It's only the `processBlockType` that does all the processing, merging the raw data, filling in the defaults etc.

**Comment by gziolo** [comment]
By the way, from my perspective, you can promote  `__experimentalReapplyBlockTypeFilters`  to a stable action. It'd be a good idea to keep an experimental name as a deprecated alias as discussed this week during WordPress Community Summit, unless folks will disagree with the notes that will get published soon at https://communitysummit.wordcamp.org/2023/.

File: packages/blocks/src/store/private-actions.js
Comment (gziolo) [inline packages/blocks/src/store/private-actions.js:34]: As far as I can tell `processBlockType` doesn't return any action to dispatch. Does it really need to be wrapped with the `dispatch` call?

**Comment by jsnajdr** [comment]
> Today, we should refactor code to use REST API endpoint for block types

This would mean that an inline script that does:
```js
wp.blocks.unstable__bootstrapServerSideBlockDefinitions({
  'core/paragraph': { title: 'Paragraph' }
});
```
would be replaced with:
```js
wp.apiFetch.use(wp.apiFetch.createPreloadingMiddleware({
  '/wp/v2/block-types': {
    'body': [ { name: 'core/paragraph', title: 'Paragraph' } ]
  }
});
```
and the access to the boostrapped block types array would be a `select` like:
```js
select( coreStore ).getEntityRecord( 'root', 'blockTypes' );
```
Is that right?

That would work, sure, there's just one caveat: `getEntityRecord` is async, and to reliably get a non-`null` value, you are _forced_ to do use the preloading middleware. You can't afford to let the data library make an actual REST request, because then `getEntityRecord` would be returning `null`, breaking the entire blocks store. In that way the REST approach is quite fragile.

> My only concern would be whether there are any legacy concerns as the post editor is initialized on the server with the context of a specific post type and a post id, so blocks in theory could get modified during registration with filters based on that information.

The preloaded REST data are created in the context of the post editor page, just like the `bootstrapServerSideBlockDefinitions` data are, so nothing really changes here.

File: packages/blocks/src/store/private-actions.js
Comment (jsnajdr) [reply packages/blocks/src/store/private-actions.js:34]: `processBlockType` is a thunk action. It can be dispatched and will return the `processedBlockType` result. The good reason to write it like this is that the thunk gets access to `select` and `dispatch` automatically.

Comment (gziolo) [reply packages/blocks/src/store/private-actions.js:34]: I figured out it could be dispatched now. I was surprised to learn it is a way to go about it. Interestingly, I shaped it in a manner compatible with `dispatch` from day one. Thank you for sharing the rationale; it is very insightful.

**Comment by gziolo** [comment]
@jsnajdr, you did a great job explaining how that could work. The REST API call would definitely need to be preloaded and further optimized so only essential fields get returned, mirroring what happens today around `wp.blocks.unstable__bootstrapServerSideBlockDefinitions` call. We can discuss it separately later.

> The preloaded REST data are created in the context of the post editor page, just like the bootstrapServerSideBlockDefinitions data are, so nothing really changes here.

Excellent observation. I didn't consider it before. In that case, it should be pretty straightforward refactoring that we can tackle after this PR lands.

File: packages/blocks/src/store/reducer.js
Comment (gziolo) [inline packages/blocks/src/store/reducer.js]: It looks like the custom logic for `apiVersion` fallback for WP 5.7 and lower can be safely removed as part of the refactoring.

File: packages/blocks/src/store/reducer.js
Comment (gziolo) [inline packages/blocks/src/store/reducer.js]: It looks like the custom logic for `ancestor` fallback for WP 5.9 and lower can be safely removed as part of the refactoring.

File: packages/blocks/src/api/registration.js
Comment (gziolo) [inline packages/blocks/src/api/registration.js]: Previously, it wouldn't run metadata processing when the first param passed to the function is the string representing the block's name. I believe it needs to remain as it as was because the `settings` object can provide a different set of keys, and some of them can be already translated like `title` or `description`. In effect, only when the `blockNameOrMetadata` is an object it should call `addBootstrappedBlock`. Historically, it was always possible to register a block without `block.json` by providing it's name and settings object that could be used as is.

File: packages/blocks/src/api/test/registration.js
Comment (gziolo) [inline packages/blocks/src/api/test/registration.js]: I think we can stabilize `__experimentalReapplyBlockTypeFilters` as part of this PR. I'm confident this action resolved most of the registration issues people dealt with in the past.

By the way, thank you so much for including a unit test covering this code in addition to the existing e2e test.

File: packages/blocks/src/store/process-block-type.js
Comment (gziolo) [inline packages/blocks/src/store/process-block-type.js:65]: Filters might modify the `blockType` object directly, so a copy `{ ...blockType }` was passed in the past. If we aren't concerned with the modifications, then maybe it would be better to use `settings` instead of the `blockType` after this line.

File: packages/blocks/src/store/private-selectors.js
Comment (gziolo) [inline packages/blocks/src/store/private-selectors.js]: This description needs to be updated to reflect its role correctly.

File: packages/blocks/src/store/private-selectors.js
Comment (gziolo) [inline packages/blocks/src/store/private-selectors.js]: Noting, that we also would use it previously for the metadata loaded from `block.json` after a light sanitization and applying translation helpers to some fields.

File: packages/blocks/src/store/private-actions.js
Comment (gziolo) [inline packages/blocks/src/store/private-actions.js]: There is some inconsistency in how `addBootstrappedBlock` and `getBootstrappedBlockType` are documented. I belive it should be both about the data registered on the server with `register_block_type` and alternatively on the client from `block.json` after light processing. I'm aware it's a bit confusing because of legacy considerations, but I'm happy to help clarify all nuances.

**Review by gziolo — COMMENTED** [review]
Excellent refactoring, @jsnajdr. I did a quite in-depth review and it's looking very close to ready. I noticed one or two potential minor issues, and some room for improvement in terms of how this code was documented in the past (that's on me) to better explain all the ways the block can get registered to satisfy different contexts: client only (covers unit tests and React Native, too) but also client + server.

**Comment by jsnajdr** [comment]
> The REST API call would definitely need to be preloaded and further optimized so only essential fields get returned, mirroring what happens today

I'm however still skeptical about whether this REST approach is the right way to go. In addition to the fragility caused by REST being async, it also adds a new `@wordpress/core-data` dependency to the `@wordpress/blocks` package. And that's undesirable. `blocks` is a WordPress-agnostic package, while `core-data` and all its registered entities are all about WordPress REST API.

File: packages/blocks/src/store/process-block-type.js
Comment (jsnajdr) [reply packages/blocks/src/store/process-block-type.js:65]: We don't need to make a copy because `blockType` is a new object (created from default, bootstrapped and unprocessed settings) on each call.

Before this PR, `processBlockType` was working directly on an object in Redux state, so a copy was needed.

Still, in both the old and new versions, there is a risk that a filter will do something like:
```js
blockType.supports.inserter = false;
```
mutating the unprocessed settings object in Redux. We'd need to make a deep copy to avoid that.

File: packages/blocks/src/store/reducer.js
Comment (jsnajdr) [reply packages/blocks/src/store/reducer.js]: Good point, removed the `apiVersion` and `ancestor` polyfills and left only the `selectors` one.

File: packages/blocks/src/api/registration.js
Comment (jsnajdr) [reply packages/blocks/src/api/registration.js]: I made this change because it's very useful for async block loading. At some places, like block inserter, my prototype looks at the server bootstrapped data, to avoid fully loading the block:
```js
const { title, parent, supports } = select( blocksStore ).getBootstrappedBlockType( 'core/query-pagination');
```

But this doesn't work for blocks that are registered only client-side, with no `block.json` metadata:
```js
registerBlockType( 'core/query-pagination', {
  title: __( 'Pagination' ),
  parent: [ 'core/query' ],
  supports: { inserter: false },
} );
```
There is no `metadata`, nothing is added to `state.bootstrappedBlockTypes` for this block. That's why I'm "reconstructing" the `metadata` by extracting selected fields from `settings`.

> the settings object can provide a different set of keys, and some of them can be already translated like `title` or `description`.

There are two scenarios that can happen here:
1. the `title` and `description` fields were bootstrapped from the server. Then they are localized (the server did that) and the call to `addBootstrappedBlock` in `registerBlockType` won't overwrite them. It will do at best the polyfilling of `.selectors`.
2. the fields were not bootstrapped from the server, and `addBoostrappedBlock` will add the data extracted from `settings` to `state.bootstrappedBlockType`. The `title` should be localized, because the JS code will call `title: __( 'Pagination' )` if the block is localized at all.

So, in neither of these cases are correct localized data overwritten with incorrent unlocalized data. We're just adding "fallback" bootstrapped metadata in case they don't exist at all.

File: packages/blocks/src/store/private-actions.js
Comment (jsnajdr) [reply packages/blocks/src/store/private-actions.js]: I improved the docblock comments for the `add*BlockType` actions and the `get*BlockType` selectors. Let me know if they need any more improvement.

I consistently talk about the "bootstrapped" data as "block type metadata" and about the "unprocessed" data as "block type settings", because they correspond exactly to the parameters of `registerBlockType( metadata, settings )`.

I also renamed the `addUnprocessedBlock` action to `addUnprocessedBlockType`, and did the same for `addBootstrappedBlock`, so that we use `*BlockType` consistently.

File: packages/blocks/src/api/test/registration.js
Comment (jsnajdr) [reply packages/blocks/src/api/test/registration.js]: Pushed a fix that stabilized this action, names it `reapplyBlockTypeFilters` and also keeps an alias with the old `__experimental` prefix, marked as deprecated :+1:

**Comment by jsnajdr** [comment]
Thanks @gziolo for a detailed review. I addressed all your feedback. The only controversial issue is whether to extract "fallback" metadata from a `registerBlockType( 'name', { ...settings } )` call. It's useful for the async block loading prototype, but it's also possible to remove this change from this PR and merge it separately, carefully tested.

File: packages/blocks/src/store/process-block-type.js
Comment (gziolo) [reply packages/blocks/src/store/process-block-type.js:65]: That is a good point about deep copy. Thinking a bit more about all of these, I think the challenge here is that block deprecations have very specific needs. You essentially want to pass the orginal `blockType` without any modifications/mutations applied with  `blocks.registerBlockType` filter to avoid the situation where the deprecation has its opinions on how to handle the same filter for every deprecation. Although the old implementation is still not resilient to mutation nested objects ...

Well, it doesn't fail any existing unit tests so that might be not an issue at all 🤷🏻

File: packages/blocks/src/api/test/registration.js
Comment (gziolo) [reply packages/blocks/src/api/test/registration.js]: Excellent, thank you so much!

File: packages/blocks/src/store/private-actions.js
Comment (gziolo) [reply packages/blocks/src/store/private-actions.js]: Sounds great, I like it a lot ❤️

File: packages/blocks/src/api/registration.js
Comment (gziolo) [reply packages/blocks/src/api/registration.js]: If it's not using `block.json`, then it should never have `textdomain` in the object, so it shouldn't run custom code that applies localization helpers. We are good in that regard.

It still isn't clear to me why it's necessary to fake the block-type bootstrapping in the case where it is registered on the client with:

> ```js
> registerBlockType( 'core/query-pagination', {
>   title: __( 'Pagination' ),
>   parent: [ 'core/query' ],
>   supports: { inserter: false },
> } );
> ```
> 
> There is no metadata, nothing is added to state.bootstrappedBlockTypes for this block. That's why I'm "reconstructing" the metadata by extracting selected fields from settings.

In this case, it should be just fine to bootstrap this block type with an empty object to signal it was also registered. Then, you could do the fallback the other way around and read values directly from the processed version of the block type. You essentially need to fully register the block type anyway to be able to expose it in the inserter.

> The only controversial issue is whether to extract "fallback" metadata from a registerBlockType( 'name', { ...settings } ) call. It's useful for the async block loading prototype, but it's also possible to remove this change from this PR and merge it separately, carefully tested.

I guess I don't fully understand the requirements for async loading, so I would say we either pass an empty object for now, or we can skip it altogether so we can discuss it separately and unblock this PR. Otherwise, I think everything is clear and I'm happy to move forward with the refactoring.

**Comment by gziolo** [comment]
> > The REST API call would definitely need to be preloaded and further optimized so only essential fields get returned, mirroring what happens today
> 
> I'm however still skeptical about whether this REST approach is the right way to go. In addition to the fragility caused by REST being async, it also adds a new `@wordpress/core-data` dependency to the `@wordpress/blocks` package. And that's undesirable. `blocks` is a WordPress-agnostic package, while `core-data` and all its registered entities are all about WordPress REST API.

We definitely don't want to make the `@wordpress/blocks` dependant on `@wordpress/core-data`. The way I was thinking about it boils down to refactoring the current approach in core where we inject the result of `get_block_editor_server_block_settings()` call with the JS version of reading a similar data from the endpoint:

https://github.com/WordPress/wordpress-develop/blob/e278fe3fcf262d1792a347a460bc004200a2490f/src/wp-admin/edit-form-blocks.php#L107-L111

Anyway, it's a separate issue, and we can discuss it later. The main reason I think it would be valuable is that today, we can only consume core blocks in the mobile app, so that would help to bridge the gap as we would use a single approach to bootstrap blocks registered on the server.

---

> The only controversial issue is whether to extract "fallback" metadata from a registerBlockType( 'name', { ...settings } ) call. It's useful for the async block loading prototype, but it's also possible to remove this change from this PR and merge it separately, carefully tested.

See my comment at https://github.com/WordPress/gutenberg/pull/53807#discussion_r1310034232.

---

It would be useful to mention two changes in the [CHANGELOG](https://github.com/WordPress/gutenberg/blob/trunk/packages/blocks/CHANGELOG.md) file:

- new stable feature: `reapplyBlockTypeFilters` action
- deprecations: `__experimentalReapplyBlockTypeFilters` action
- should we also mention the removed experiments that are now private APIs?

File: packages/blocks/src/api/registration.js
Comment (jsnajdr) [reply packages/blocks/src/api/registration.js]: I reverted to the old boostrapping behavior in e6dae48d81de6c404e2c301141606209ad0bfa00. Extracting the `metadata` from `settings` can be done later, and can be discussed with better context about async block loading.

Having done this, I believe all feedback is addressed :+1:

File: packages/blocks/src/store/reducer.js
Comment (jsnajdr) [inline packages/blocks/src/store/reducer.js:84]: Added this comment, and it implements the suggestion I shared with @ockham in https://github.com/WordPress/gutenberg/pull/52969/files#r1310187507.

**Review by gziolo — APPROVED** [review]
Let's go ahead and move forward now that all feedback has been taken care of. Thank you so much for walking me through all the changes proposed. I'm delighted to see all the improvements applied.

**Pull request merged**
